src/Workspaces/Remote/Core/ProjectSystem/IWorkspaceProjectFactoryService.cs

```diff
@@ -0,0 +1,25 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.CodeAnalysis.Remote.ProjectSystem
+{
+    internal interface IWorkspaceProjectFactoryService
+    {
+        Task<IWorkspaceProject> CreateAndAddProjectAsync(WorkspaceProjectCreationInfo creationInfo, CancellationToken cancellationToken);
+
+        /// <summary>
+        /// Returns the list of properties that are understood by the language service and can be passed to
+        /// <see cref="IWorkspaceProject.SetBuildSystemPropertiesAsync(IReadOnlyDictionary{string, string}, CancellationToken)"/> and to
+        /// <see cref="WorkspaceProjectCreationInfo.BuildSystemProperties"/>.
+        /// </summary>
+        Task<IReadOnlyCollection<string>> GetSupportedBuildSystemPropertiesAsync(CancellationToken cancellationToken);
```

Comment:
Note that the project system may not be able to support arbitrary property retrieval in the future. We do have that support at the moment but only through a very atypical use of MSBuild/CPS APIs that may not be available in the future. The idiomatic approach in CPS is to export a `Rule` object that defines ahead of time which properties/items/metadata need to be harvested from the results of an evaluation or design-time build.

We should probably discuss that further outside of the code review. The immediate implication is that I don't expect the project system will ever call this method.

Reply:
This is currently implemented via `IWorkspaceProjectContextFactory.EvaluationPropertyNames` and `IWorkspaceProjectContextFactory.EvaluationItemNames`, so I assume this method is the equivallent. The project system can call these ahead of time and construct the rule objects as needed.

Reply:
Yes, this is method is intended as the replacement for EvaluationPropertyNames; we don't have an equivalent for EvaluationItemNames since that's going to need more design work for now.

Reply:
Further discussion with @tmeschter meant he wasn't entirely sure what the implementation to consume this would look like, since CPS introduces a bit of a catch-22 where they'd have to block on this call fairly early during their load process to actually create the rules before the project system then runs.

Note as it stands right now the project system doesn't actually use EvaluationPropertyNames anywhere.

At this point, I'm going to leave this as is. If @tmeschter can consume this without too much difficulty, then I think that'll give us further confidence to also flesh back out the item support. If this doesn't work, then we'll remove this with a different approach once we know understand our new constraints.